### ISSUE TITLE
ODataNextLink When Calling Drives["driveid"].Items returns a url for Sharepoint Version 2 instead of Graph, causing an invalid audience exception

### ISSUE THREAD

**Issue by ryanwilson-gpdgroup** [microsoftgraph/msgraph-sdk-dotnet#2316]
**Describe the bug**

I make a request to Drives.Items and receive a DriveItemCollectionResponse. The OdataNextLink then gives me a Url that points to sharepoint.com/_api/v2.0. Since my audience on my JWT is for graph.microsoft.com it fails on the request to get the next page of data.
 
Produces:

**Exception: Microsoft.Graph.Models.ODataErrors.ODataError: Invalid audience Uri 'https://graph.microsoft.com/'.**

**To Reproduce**
_When using the GraphServiceClient class I make a request for Items of a drive like so, with the base uri in the initial request showing as: https://graph.microsoft.com/v1.0_

```
DriveItemCollectionResponse drivecontent = await _graphclient.Drives["somedriveid"].Items.GetAsync(rc =>
{
                rc.QueryParameters.Select = new string[] 
                {
                    "name",
                    "id",
                    "description"
                };
                rc.QueryParameters.Filter = "Name ne ' '";
 });
```

The OdataNextLink has a url then for _"https://xxxxx.sharepoint.com/_api/v2.0/"_, so on my next call using:
```
var results = await _graphclient.Drives[$"somedriveid"].Items.WithUrl(drivecontent.OdataNextLine).GetAsync();
```
**I get the above exception.**

**AND my Token Provider**
 ```
public class GraphTokenProvider : IAccessTokenProvider
 {
     private readonly IConfiguration _configuration;
     
     public GraphTokenProvider(IConfiguration configuration)
     {
         _configuration = configuration;
     }

     public async Task<string> GetAuthorizationTokenAsync(
         Uri uri, Dictionary<string, object> additionalAuthenticationContext = default,
     CancellationToken cancellationToken = default)
     {
         IConfidentialClientApplication confidentialClientApplication = ConfidentialClientApplicationBuilder
             .Create(_configuration.GetValue<string>("AzureAd:ClientId"))
             .WithTenantId(_configuration.GetValue<string>("AzureAd:TenantId"))
             .WithClientSecret(_configuration.GetValue<string>("AzureAd:ClientSecret"))
             .Build();
         // The app registration should be configured to require access to permissions
         // sufficient for the Microsoft Graph API calls the app will be making, and
         // those permissions should be granted by a tenant administrator.
         var scopes = new string[] { "https://graph.microsoft.com/.default" };

         var token = await confidentialClientApplication.AcquireTokenForClient(scopes).ExecuteAsync();
         // get the token and return it in your own way
         return token.AccessToken;
     }

     public AllowedHostsValidator AllowedHostsValidator { get; }
 }
```

I can get a successful paged response if I replace:  xxxxx.sharepoint.com/_api/v2.0/ with graph.microsoft.com/v1.0 in the OdataNextLink string value but the response returns results from the first page and not a correct paged result.

**Expected behavior**

OdataNextLink should be a Url targeting graph.microsoft.com

**Screenshots**
n/a

**Client version**
Microsoft.Graph 5.40.0

**Desktop (please complete the following information):**
N/A

**Additional context**

As this may be a problem on Microsoft's end, I apologize if I should have opened this somewhere else. If so, can someone help me to submit this to the appropriate party?

**Comment by mewanZCW** [comment]
@ryanwilson-gpdgroup  I also had the same issue. already raised a issue here #2291. Still waiting for a fix

**Comment by ryanwilson-gpdgroup** [comment]
@mewanZCW Thanks for the information. I'll give that a look.

**Comment by ryanwilson-gpdgroup** [comment]
@mewanZCW  I used the following:

```
DriveItemCollectionResponse drivecontent = await _graphclient.Drives[$"{driveId}"].Items.GetAsync(rc =>
{
    rc.QueryParameters.Select = new string[] 
    {
        "name",
        "id",
        "description"
    };
    rc.QueryParameters.Filter = "Name ne ' '";
});


var pageIterator = PageIterator<DriveItem, DriveItemCollectionResponse>
    .CreatePageIterator((IBaseClient)_graphclient, drivecontent,callback:
    (rc) =>
    {
        drivecontent.Value?.Add(rc);
        return true;
    });

await pageIterator.IterateAsync();
```

And it produced the same issue with the audience being invalid on the token.

**Comment by mewanZCW** [comment]
@ryanwilson-gpdgroup  I guess this is not related to a token issue. If you want to make sure you can generate a token from Graph Explorer and try with that. Looks like this issue is having with the drives. If you use my previous example with expanding the drive you'll still get the drive details and you can do the filtering manually. But you need to know the site and and list/libraryId to use with that. 

this was my old code with that error.
```
List<DriveItem> childrens = new List<DriveItem>();
DriveItemCollectionResponse response = new DriveItemCollectionResponse();
if (isRoot)
{
    response = await _graphClient.Drives[driveId].Items.GetAsync((requestConfiguration) =>
    {
        requestConfiguration.QueryParameters.Filter = $"parentReference/path eq '/drives/{driveId}/root:'";
    });
}

else
{
    response = await _graphClient.Drives[driveId].Root.ItemWithPath(folderPath).Children.GetAsync();
}

var pageIterator = PageIterator<DriveItem, DriveItemCollectionResponse>
    .CreatePageIterator(_graphClient, response,
    (children) =>
    {
        childrens.Add(children);
        return true;
    });

await pageIterator.IterateAsync();

childrens = childrens.Where(x => x.Folder != null).ToList();
return childrens;
```

and this is the alternative solution that worked for me 

```
ListItemCollectionResponse response = new ListItemCollectionResponse();

response = await _graphClient.Sites[siteId].Lists[listId].Items.GetAsync((requestConfiguration) =>
{
    requestConfiguration.QueryParameters.Expand = new string[] { "DriveItem" };
});

var pageIterator = PageIterator<ListItem, ListItemCollectionResponse>
    .CreatePageIterator(_graphClient, response,
    (children) =>
    {
        childrens.Add(children);
        return true;
    });

await pageIterator.IterateAsync();

if (isRoot)
{
    childrens = childrens.Where(x => x.DriveItem.ParentReference.Path == $"/drives/{driveId}/root:").ToList();
}
else
{
    childrens = childrens.Where(x => x.DriveItem.ParentReference.Path == $"/drives/{driveId}/root:/{folderPath}").ToList();
}

childrens = childrens.Where(x => x.DriveItem.Folder == null).ToList();
```

Hope you can get a solution by looking at this code. It should work if you go with listItems not that Drive.

**Comment by MartinM85** [comment]
It's not related to SDK but to the Graph API itself.

Usually `https://graph.microsoft.com/v1.0/drives/{drive_id}/items` is not officially supported to retrieve all items, but it works with the filter, which is cool.

From a quick test in my tenant `https://graph.microsoft.com/v1.0/drives/{drive_id}/items?$filter=Name ne ' '` doesn't return all items, so is not 100% reliable.

Not sure, what are you trying to achieve? Find some file or get the whole structure of a some drive?

**Comment by ryanwilson-gpdgroup** [comment]
@MartinM85 I am basically trying to get the name/id properties of every item contained within a drive (folders,files etc.). Thank you for your response.

**Comment by fzbm** [comment]
It looks like the issue has just been fixed. 👍

I was in the process to implement the URL replacement workaround and realized, that the `nextLink` always contained a Graph URL. Additional tests via the Graph Explorer and reviewing production logs confirmed my assumption. I performed the same Graph Explorer tests about an hour ago and was still retrieving the SharePoint URLs.

The affected Graph request was `/me/drive/following`.
